### DISCUSSION TITLE
Max throughput of USB Serial (CDC) over USB High Speed

### DISCUSSION THREAD

**Discussion by cbrake** [hathach/tinyusb#1659] [Q&A] [answered]
Hi, does anyone know of the max throughput (rough numbers are fine) you can obtain over USB Serial (CDC) on USB High Speed? We are using a STM32H7 MCU. Are we limited to classic UART baud rates like 115200, or is BAUD rate ignored for USB serial when the endpoint is a MCU?

**Comment by manuelbl** [comment, marked as answer]
Baud rate is ignored. So in theory, you should get close to the speed limit of high speed USB, i.e. about 30 to 40 MByte/s.

Note however that it depends on your operating system's implementation of serial communication. It might not be optimized for such high speeds as the typical serial speeds max out at about 0.1 MByte/s.

Why do you want to use CDC and not a vendor specific protocol?

**Comment by cbrake** [comment]
I've been doing [some testing](https://github.com/simpleiot/firmware/tree/master/Arduino) on a SAMD21 running Arduino SW, and getting around 2.4Mb/sec over a 12Mb full speed link over CDC, so that seems pretty good! Another team member is still developing the tinyusb based solution on a STM32 with high-speed USB.

In this application, I prefer CDC as I don't need to mess with drivers on the host side. Using COBS, packets can be framed, and nanopb is working well for encoding. (see above example).

Open to other ideas as I don't have a lot of USB experience!

**Reply by manuelbl** [reply]
What's your host hardware and operating system? A Raspberry Pi?

CDC (or a serial connection) is a valid option if it is not a consumer application as the user is usually forced to figure out which of the multiple serial connection available on the host is the one for the attached device.

For communication, both CDC and a vendor interface use bulk endpoints. So unless there is an inefficiency in the host's serial communication drivers, the performance should be similar.

I'm not familiar with COBS. But nanopb can also be used with the vendor specific interface. That's also similar.

**Reply by cbrake** [reply]
> What's your host hardware and operating system? A Raspberry Pi?

Pretty much any Windows/Mac/Linux system: https://github.com/simpleiot/simpleiot/releases

> 
> CDC (or a serial connection) is a valid option if it is not a consumer application as the user is usually forced to figure out which of the multiple serial connection available on the host is the one for the attached device.

That is a good point -- one of the tradeoffs. We might eventually provide a way to identify the device using USB IDs rather than serial port, which would make it a little more reliable.

> For communication, both CDC and a vendor interface use bulk endpoints. So unless there is an inefficiency in the host's serial communication drivers, the performance should be similar.

That is good to know!

> I'm not familiar with COBS. But nanopb can also be used with the vendor specific interface. That's also similar.

This is a good overview of COBS: http://www.stuartcheshire.org/papers/COBSforToN.pdf

COBS is a good way to delineate packets in a serial stream.

I looked up a little about the vendor class -- how do you typically implement host drivers for this? I see there is WebUSB ...

**Reply by manuelbl** [reply]
It looks as if your host code is written in Go. Thus you'd probably want to use a package like *gousb* as the native API on Windows, Linux and Mac are totally different. Communication looks like in this example: https://pkg.go.dev/github.com/google/gousb#example-package-Simple, i.e. get the USB device by VID/PID, open it, claim a USB interface and start sending and receiving data.

Like many USB libraries, *gousb* depends on *libusb*. I don't know if your software provides an installer. If not, the *libusb* installation can be a challenge for users. You will likely run into the same problems if you try to identify the serial device by USB IDs because most of the libraries that can do that also depend in *libusb*.

**Reply by cbrake** [reply]
yeah, I'd really like to avoid libusb and stay pure Go, so for this type of functionality (VID/PID) I may focus on Linux for now, and implement a pure Go version.

**Reply by archef2000** [reply]
Have you ever achieved more than 2.4 Mbps over the serial port with TinyUSB?
